pkg/k8s/rule_translate.go

```diff
@@ -134,21 +134,26 @@ func generateToCidrFromEndpoint(
 		}
 
 		found := false
-		for _, c := range egress.ToCIDRSet {
+		for i, c := range egress.ToCIDRSet {
 			_, cidr, err := net.ParseCIDR(string(c.Cidr))
 			if err != nil {
 				return err
 			}
-			if cidr.Contains(epIP) {
+			// Add a generated CIDR even if a non-generated CIDR might already contain 'epIP'.
+			// Also check that the IP in cidr is the same as 'epIP', in case we have a
+			// generated CIDR with shorter prefix containing 'epIP'.
+			if c.Generated > 0 && cidr.Contains(epIP) && cidr.IP.Equal(epIP) {
```

Comment:
This is a little odd to me. What happens if we have 2 backends {`10.0.0.1` and `10.0.0.2`}. The first `epIP(10.0.0.1)` will generate a `ToCIDRSet` with `Cidr: 10.0.0.0/24`. The second `epIP(10.0.0.2)` will also generate a `ToCIDRSet` because `cidr.IP.Equal(epIP)` is false since `cidr.IP` is `10.0.0.0`. Is my understanding correct? Is this the intended behavior?

Reply:
Given this snippet earlier in the function, the CIDR for `10.0.0.1` will be `10.0.0.1/32`, not `10.0.0.0/24`?
```
	// This will generate one-address CIDRs consisting of endpoint backend ip
	mask := net.CIDRMask(128, 128)
```
So each backend will have their own CIDR. Also, these *one-address* CIDRs will now be generated even if `ToCIDRSet` already contained a non-generated CIDR that contains the one-address CIDR so that we have an accurate `Generated` count for each generated CIDR.

Currently all generated CIDRs are of this *one-address* CIDR type (I think), so strictly speaking the added `Equal` check is not necessary, but exists for the case that `ToCIDRSet` may contain wider generated CIDRs at a later time.

Reply:
> So each backend will have their own CIDR. Also, these one-address CIDRs will now be generated even if ToCIDRSet already contained a non-generated CIDR that contains the one-address CIDR so that we have an accurate Generated count for each generated CIDR.

Is it possible to mix ToCIDR and ToServices fields into the same EgressRule from the API? Or is this only possible internally when Cilium populates the ToCIDR fields in a ToServices rule per this logic?

> strictly speaking the added Equal check is not necessary, but exists for the case that ToCIDRSet may contain wider generated CIDRs at a later time.

I'm a bit skeptical that we could somehow reuse this logic for wider CIDR prefixes without rethinking the logic. Personally I'd be inclined to avoid any code that suggests that case is supported unless we're actually generating & testing such CIDR prefixes here.

Reply:
@jrajahalme you are right. I missed the `mask` above. I'll leave this unresolved so that Joe's comments don't get lost.